library/hatomap.py

```diff
@@ -0,0 +1,478 @@
+            pixel_x_west=origin_pixel_x,
+            pixel_x_east=origin_pixel_x + self.width,
+            pixel_y_north=origin_pixel_y,
+            pixel_y_south=origin_pixel_y + self.height,
+            zoom=self.zoom,
+        )
+
+
+@dataclass
+class RasterTileServer:
+    url: str
+
+    @staticmethod
+    def _get_image_content(url):
+        return cv2.imdecode(
+            np.asarray(bytearray(requests.get(url).content), dtype=np.uint8), -1
+        )
+
+    def request(self, bbox: WebMercatorPixelBBox = None) -> np.ndarray:
+        if bbox is None:
+            return None
+
+        (tl_tilepx, rb_tilepx) = bbox.covered_tiles()
+
+        request_urls = []
+        for x in range(tl_tilepx.tile.tile_x, rb_tilepx.tile.tile_x + 1):
+            for y in range(tl_tilepx.tile.tile_y, rb_tilepx.tile.tile_y + 1):
+                request_urls.append(
+                    string.Template(self.url).safe_substitute(
+                        {"x": x, "y": y, "z": bbox.zoom}
+                    )
+                )
+        with Pool(16) as p:
+            imgs = list(p.imap(self._get_image_content, request_urls))
+
+        tile_width_cnt = rb_tilepx.tile.tile_x + 1 - tl_tilepx.tile.tile_x
+        tile_height_cnt = rb_tilepx.tile.tile_y + 1 - tl_tilepx.tile.tile_y
+
+        concated = np.concatenate(
+            [
+                np.concatenate(imgs[i:i + tile_height_cnt], axis=0) for i in range(0, len(imgs), tile_height_cnt)
+            ],
+            axis=1,
+        )
+
+        px_top = tl_tilepx.tile_pixel_y
+        px_bottom = (tile_height_cnt - 1) * 256 + rb_tilepx.tile_pixel_y
+        px_left = tl_tilepx.tile_pixel_x
+        px_right = (tile_width_cnt - 1) * 256 + rb_tilepx.tile_pixel_x
+        trim = concated[px_top:px_bottom, px_left:px_right]
+        return trim
+
+
+@dataclass
+class LineTrace:
+    coords: Union[List["np.ndarray"], List[List[GeoCoord]]]
+    width: int = 1
+    color: Tuple[int, int, int, int] = (255, 0, 255, 255)
+
+    def get_image(self, bbox: WebMercatorPixelBBox = None) -> np.ndarray:
+        img = np.zeros((bbox.height, bbox.width, 4), np.uint8)
+        if type(self.coords[0][0]) is GeoCoord:
+            coords = [
+                np.array([[g.lat, g.lng] for g in coords]) for coords in self.coords
+            ]
+        elif type(self.coords[0][0][0]) is np.float64:
+            coords = self.coords
+        else:
+            raise TypeError("coords should be List[np.ndarray] or List[List[GeoCoord]]")
+
+        px_coords = bbox.geocoords2pixel(coords)
+        cv2.polylines(
+            img,
+            px_coords,
+            color=self.color,
+            thickness=self.width,
+            isClosed=True,
+            lineType=cv2.LINE_AA,
+        )
+        return img
+
+
+@dataclass
+class MarkerTrace:
+    coords: Union["np.ndarray", List[GeoCoord]]
+    symbol: str = "circle"
+    size: int = 4
+    border_color: Tuple[int, int, int] = (255, 0, 255, 255)
```

Comment:
```suggestion
    border_color: Tuple[int, int, int, int] = (255, 0, 255, 255)
```

https://github.com/dev-hato/hato-bot/actions/runs/2783525807/jobs/4380588635

```
library/hatomap.py:252: error: Incompatible types in assignment (expression has type "Tuple[int, int, int, int]", variable has type "Tuple[int, int, int]")
```

`Tuple` 内の要素数が合ってないね。